// @appscode/design-system/plugins/vue-toaster.js
module.exports = {
  position: "top-center",
  theme: "toasted-primary",
  iconPack: "fontawesome",
  className: "ac-toast",
  action: {
    icon: "close",
    class: "ac-toast-action",
    onClick: (e, toastObject) => {
      toastObject.goAway(0);
    },
  },
  register: [
    {
      name: "success",
      message: (payload) => {
        if (payload) {
          if (typeof payload === "string") return payload;
          else if (payload.message) return payload.message;
          else return "Operation Completed Successfully";
        }
        // if there is no message passed show default message
        else return "Operation Completed Successfully";
      },
      options: {
        icon: "fa-check",
        duration: 2500,
        type: "success",
        className: "ac-toast is-success",
      },
    },
    {
      name: "error",
      message: (payload) => {
        if (payload) {
          if (typeof payload === "string") return payload;
          else if (payload.response && payload.response.data && payload.response.data.message)
            return payload.response.data.message;
          else if (payload.message) return payload.message;
          else return "Something Went Wrong";
        } else return "Something Went Wrong";
      },
      options: {
        icon: "fa-warning",
        duration: 20000,
        type: "error",
        className: "ac-toast is-danger",
      },
    },
    {
      name: "info",
      message: (payload) => {
        if (payload) {
          if (typeof payload === "string") return payload;
          else if (payload.message) return payload.message;
          else return "Something Went Wrong";
        } else return "Something Went Wrong";
      },
      options: {
        icon: "fa-info",
        duration: 20000,
        type: "info",
        className: "ac-toast is-info",
      },
    },
    {
      name: "warning",
      message: (payload) => {
        if (payload) {
          if (typeof payload === "string") return payload;
          else if (payload.message) return payload.message;
          else return "Something Went Wrong";
        }
        // if there is no message passed show default message
        else return "Something Went Wrong";
      },
      options: {
        icon: "fa-warning",
        duration: 2500,
        type: "error",
        className: "ac-toast is-warning",
      },
    },
  ],
};
